sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -3805,3 +3799,330 @@ object ArrayUnion {
+           |  } else {
+           |    $hs.add$postFix($array2.$getter);
+           |  }
+           |}
+           |for (int $i = 0; $i < $array1.numElements(); $i++) {
+           |  if ($array1.isNullAt($i)) {
+           |    if ($notFoundNullElement) {
+           |      $size++;
+           |      $notFoundNullElement = false;
+           |    }
+           |  } else {
+           |    $javaTypeName $value = $array1.$getter;
+           |    if (!$hs.contains($castOp $value)) {
+           |      $hs.add$postFix($value);
+           |      $size++;
+           |    }
+           |  }
+           |}
+           |$arrayBuilder
+           |$hs = new $openHashSet$postFix($classTag);
+           |$notFoundNullElement = true;
+           |int $pos = 0;
+           |for (int $i = 0; $i < $array2.numElements(); $i++) {
+           |  if ($array2.isNullAt($i)) {
+           |    $notFoundNullElement = false;
+           |  } else {
+           |    $hs.add$postFix($array2.$getter);
+           |  }
+           |}
+           |for (int $i = 0; $i < $array1.numElements(); $i++) {
+           |  if ($array1.isNullAt($i)) {
+           |    if ($notFoundNullElement) {
+           |      ${ev.value}.setNullAt($pos++);
+           |      $notFoundNullElement = false;
+           |    }
+           |  } else {
+           |    $javaTypeName $value = $array1.$getter;
+           |    if (!$hs.contains($castOp $value)) {
+           |      $hs.add$postFix($value);
+           |      ${ev.value}.$setter;
+           |      $pos++;
+           |    }
+           |  }
+           |}
+         """.stripMargin
+      } else {
+        val arrayExcept = classOf[ArrayExcept].getName
+        val et = ctx.addReferenceObj("elementTypeIntersect", elementType)
+        val order = ctx.addReferenceObj("orderingIntersect", ordering)
+        val method = "exceptOrdering"
+        s"${ev.value} = $arrayExcept$$.MODULE$$.$method($array1, $array2, $et, $order);"
+      }
+    })
+  }
+
+  override def prettyName: String = "array_except"
+}
+
+object ArrayExcept {
+  def exceptOrdering(
+      array1: ArrayData,
+      array2: ArrayData,
+      elementType: DataType,
+      ordering: Ordering[Any]): ArrayData = {
+    val arrayBuffer = new scala.collection.mutable.ArrayBuffer[Any]
+    var scannedNullElements = false
+    var i = 0
+    while (i < array1.numElements()) {
+      var found = false
+      var elem1 = array1.get(i, elementType)
+      if (array1.isNullAt(i)) {
+        if (!scannedNullElements) {
+          var j = 0
+          while (!found && j < array2.numElements()) {
+            found = array2.isNullAt(j)
+            j += 1
+          }
+          // array2 is scanned only once for null element
+          scannedNullElements = true
+        } else {
+          found = true
+        }
+      } else {
+        var j = 0
+        while (!found && j < array2.numElements()) {
+          if (!array2.isNullAt(j)) {
```

Comment:
nit: Couldn't we shorten the null check as on the line [4116](https://github.com/apache/spark/pull/21103/files#diff-9853dcf5ce3d2ac1e94d473197ff5768R4116)?